roles/lib_openshift/src/test/unit/test_oc_serviceaccount_secret.py

```diff
@@ -1,414 +1,199 @@
+    ])
+
+    # assert_called_once is not available in python 3.5
+    assert spy_write.call_count == 1
+    assert yaml_provider.safe_load(spy_write.call_args[0][1]) == builder
+
+
+def test_removing_a_secret_to_a_serviceaccount(yaml_provider, mocker, service_account, builder):
+    ''' Testing removing a secret to a service account '''
+    mocker.patch('oc_serviceaccount_secret.locate_oc_binary', return_value='oc')
+    mocker.patch('oc_serviceaccount_secret.Utils.create_tmpfile_copy', return_value='/tmp/mocked_kubeconfig')
+    spy_write = mocker.spy(oc_serviceaccount_secret.Yedit, '_write')
+
+    # Arrange
+
+    # run_ansible input parameters
+    params = {
+        'state': 'absent',
+        'namespace': 'default',
+        'secret': 'newsecret',
+        'service_account': 'builder',
+        'kubeconfig': '/etc/origin/master/admin.kubeconfig',
+        'debug': False,
+    }
+
+    new_secret = {'name': 'newsecret'}
+
+    service_account_before = copy.deepcopy(service_account)
+    service_account_before['secrets'].append(new_secret)
+    oc_get_sa_before = json.dumps(service_account_before)
+
+    # Return values of our mocked function call. These get returned once per call.
+    mock_cmd = mocker.patch('oc_serviceaccount_secret.OCServiceAccountSecret._run',
+                            side_effect=[
+                                (0, oc_get_sa_before, ''),  # First call to the mock
+                                (0, oc_get_sa_before, ''),  # Second call to the mock
+                                (0, 'serviceaccount "builder" replaced', ''),  # Third call to the mock
+                            ])
+
+    # Act
+    results = oc_serviceaccount_secret.OCServiceAccountSecret.run_ansible(params, False)
+
+    # Assert
+    assert results['changed'] is True
+    assert results['results']['returncode'] == 0
+    assert results['state'] == 'absent'
+
+    # Making sure our mocks were called as we expected
+    mock_cmd.assert_has_calls([
+        mock.call(['oc', 'get', 'sa', 'builder', '-o', 'json', '-n', 'default'], None),
+        mock.call(['oc', 'get', 'sa', 'builder', '-o', 'json', '-n', 'default'], None),
+        mock.call(['oc', 'replace', '-f', mock.ANY, '-n', 'default'], None),
+    ])
+
+    # assert_called_once is not available in python 3.5
+    assert spy_write.call_count == 1
+    assert yaml_provider.safe_load(spy_write.call_args[0][1]) == builder
+
+
+def test_binary_lookup(binary_lookup_test_data, mocker):
+    path = binary_lookup_test_data['path']
+    which_result = binary_lookup_test_data['which_result']
+    binary_expected = binary_lookup_test_data['binary_expected']
+    binaries = binary_lookup_test_data['binaries']
+
+    mocker.patch('os.environ.get', side_effect=lambda _v, _d: path)
+
+    if six.PY2:
+        mocker.patch('os.path.exists', side_effect=lambda f: f in binaries)
+    else:
+        mocker.patch('shutil.which', side_effect=lambda _f, path=None: which_result)
```

Comment:
In this particular case, both `os.environ.get` and `os.path.exists` are only needed in PY2.

I'd suggest the simplification to always patch everything, without the distinction of Python version. `shutil.which` is an exception because it is not defined in PY2:

```python
def test_binary_lookup(binary_lookup_test_data, mocker):
    path = binary_lookup_test_data['path']
    binaries = binary_lookup_test_data['binaries']
    which_result = binary_lookup_test_data['which_result']
    binary_expected = binary_lookup_test_data['binary_expected']

    mocker.patch('os.environ.get', side_effect=lambda _v, _d: path)
    mocker.patch('os.path.exists', side_effect=lambda f: f in binaries)

    if six.PY3:
        mocker.patch('shutil.which', side_effect=lambda _f, path=None: which_result)

    assert oc_serviceaccount_secret.locate_oc_binary() == binary_expected
```